plugins/workspace-plugin/src/workspace-folder-updater.ts

```diff
@@ -0,0 +1,91 @@
+/**********************************************************************
+ * Copyright (c) 2021 Red Hat, Inc.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ***********************************************************************/
+
+import * as theia from '@theia/plugin';
+const UPDATE_WORKSPACE_FOLDER_TIMEOUT = 5000;
+
+export class WorkspaceFolderUpdater {
+  private pendingFolders: string[] = [];
+  private addingWorkspaceFolderPromise: Promise<void> | undefined;
+
+  async addWorkspaceFolder(path: string): Promise<void> {
+    const workspaceFolderPath = this.toValidWorkspaceFolderPath(path);
+    if (this.pendingFolders.includes(workspaceFolderPath)) {
+      return Promise.resolve();
+    }
+
+    if (this.addingWorkspaceFolderPromise) {
+      this.pendingFolders.push(workspaceFolderPath);
+    } else {
+      try {
+        this.addingWorkspaceFolderPromise = this.addFolder(workspaceFolderPath);
+        await this.addingWorkspaceFolderPromise;
+      } catch (error) {
+        console.error(error);
+      } finally {
+        this.addingWorkspaceFolderPromise = undefined;
+      }
+
+      const next = this.pendingFolders.shift();
+      if (next) {
+        this.addWorkspaceFolder(next);
+      }
+    }
+    return Promise.resolve();
+  }
+
+  protected addFolder(projectPath: string): Promise<void> {
+    const workspaceFolders = theia.workspace.workspaceFolders;
+    const workspaceFoldersPaths = workspaceFolders ? workspaceFolders.map(folder => folder.uri.path) : [];
+    if (workspaceFoldersPaths.includes(projectPath)) {
+      return Promise.resolve(undefined);
+    }
+
+    return new Promise<void>((resolve, reject) => {
+      const disposable = theia.workspace.onDidChangeWorkspaceFolders(event => {
+        const existedWrokspaceFolders = theia.workspace.workspaceFolders;
+
+        const existedFoldersPaths = existedWrokspaceFolders
+          ? existedWrokspaceFolders.map(folder => folder.uri.path)
+          : [];
+        const addedFoldersPaths = event.added.map(folder => folder.uri.path);
+
+        if (addedFoldersPaths.includes(projectPath) || existedFoldersPaths.includes(projectPath)) {
```

Comment:
Why would we check the existing folders? The existing folders may be updated before the event is sent and we have no guarantee that processing is complete just because the folder is in "existingFolders". Why not just check for the added folders? Is the event not reliable?

Reply:
Let's consider the case - we subscribed on the event too late - the folder was already added
If we don't check `existingWorkspaceFolders` - we never find the folder among `added` folders as we already missed the corresponding event. We can find it only among `existingWorkspaceFolders`.  

```
existingWorkspaceFolders = theia.workspace.workspaceFolders
```
so - we should do nothing - the folder was already added